core/src/main/java/com/linecorp/armeria/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.concurrent.FastThreadLocal;
+
+/**
+ * A {@link RequestContextStorage} which keeps track of {@link RequestContext}s, reporting pushed thread
+ * information if a {@link RequestContext} is leaked.
+ */
+public final class LeakTracingRequestContextStorage implements RequestContextStorage {
+
+    private final RequestContextStorage delegate;
+    private final FastThreadLocal<PendingRequestContextStackTrace> pendingRequestCtx;
+    private final Sampler<Class<? extends Throwable>> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate) {
+        this(delegate, Flags.verboseExceptionSampler());
+    }
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                            Sampler<Class<? extends Throwable>> sampler) {
+        this.delegate = delegate;
+        pendingRequestCtx = new FastThreadLocal<>();
+        this.sampler = sampler;
```

Comment:
I don't see we need to limit the type to `Class<Throwable>`.
```suggestion
                                            Sampler<?> sampler) {
        this.delegate = delegate;
        pendingRequestCtx = new FastThreadLocal<>();
        this.sampler = (Sampler<Object>) sampler;
```